crates/flipperzero/src/gpio/i2c.rs

```diff
@@ -0,0 +1,402 @@
+    pub fn is_device_ready(&mut self, device: DeviceAddress, timeout: u32) -> bool {
+        unsafe { sys::furi_hal_i2c_is_device_ready(self.handle, device.0, timeout) }
+    }
+
+    /// Reads the 8-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn read_u8(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        timeout: u32,
+    ) -> Result<u8, Error> {
+        let mut data = 0;
+        if unsafe {
+            sys::furi_hal_i2c_read_reg_8(self.handle, device.0, reg_addr, &mut data, timeout)
+        } {
+            Ok(data)
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Reads the 16-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn read_u16(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        timeout: u32,
+    ) -> Result<u16, Error> {
+        let mut data = 0;
+        if unsafe {
+            sys::furi_hal_i2c_read_reg_16(self.handle, device.0, reg_addr, &mut data, timeout)
+        } {
+            Ok(data)
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Reads `device`'s memory starting at `mem_addr` into the given buffer.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn read_exact(
+        &mut self,
+        device: DeviceAddress,
+        mem_addr: u8,
+        buf: &mut [u8],
+        timeout: u32,
+    ) -> Result<(), Error> {
+        if unsafe {
+            sys::furi_hal_i2c_read_mem(
+                self.handle,
+                device.0,
+                mem_addr,
+                buf.as_mut_ptr(),
+                buf.len(),
+                timeout,
+            )
+        } {
+            Ok(())
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Writes the given value into the 8-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn write_u8(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        data: u8,
+        timeout: u32,
+    ) -> Result<(), Error> {
+        if unsafe { sys::furi_hal_i2c_write_reg_8(self.handle, device.0, reg_addr, data, timeout) }
+        {
+            Ok(())
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Writes the given value into the 16-bit register at `reg_addr` on `device`.
+    ///
+    /// `timeout` is in milliseconds.
+    pub fn write_u16(
+        &mut self,
+        device: DeviceAddress,
+        reg_addr: u8,
+        data: u16,
+        timeout: u32,
+    ) -> Result<(), Error> {
+        if unsafe { sys::furi_hal_i2c_write_reg_16(self.handle, device.0, reg_addr, data, timeout) }
+        {
+            Ok(())
+        } else {
+            Err(Error::TransferFailed)
+        }
+    }
+
+    /// Writes the given data into `device`'s memory starting at `mem_addr`.
```

Comment:
I might be missing something, but it looks like `Bus` provides almost exactly the same methods as `BusHandle`.

It makes it a bit tricky to tell what is the correct way to use the API. I would normally expect an API that looked something like the following:

```rust

const i2c::Bus::INTERNAL: &mut sys::FuriHalI2cBusHandle = &mut sys::furi_hal_i2c_handle_power;

{
    let bus = i2c::Bus::acquire(i2c::Bus::INTERNAL);

    if bus.is_device_ready(i2c::DeviceAddress::new(0x68), 0) {
        ...
    }

    // Bus is released when dropped
}
```

Reply:
That's what I originally did, but then I looked into how I2C was used elsewhere by Flipper Zero, and it seemed like only acquiring as long as you need it was the way to go sometimes? But then I figured some issue would want to acquire for longer periods. So the `Bus` methods are just simple wrappers when you only need to call something once.

All this is to say, I'd love some input from someone who works with I2C more than I have (never).

Reply:
I've removed the convenience methods from `i2c::Bus` to simplify this PR. Users can do the same thing directly by calling `i2c::Bus::with_handle`.